### PR TITLE
core: store and check for Raft version changes

### LINKED ISSUE #11867: Server with raft protocol 2 joining cluster with raft protocol 3
### Nomad version
1.2.3

### Operating system and Environment details
CentOS Stream 8

### Issue
In my configuration, I set `raft_protocol` to 3 some time ago to upgrade all my servers. Now that I'm upgrading to 1.2.3 I wanted to remove that hardcoded version in case at some point there's a version 4. I removed it from one server and I was surprised to see it rejoin the cluster with a protocol of 2:

```
Node       ID                 Address      State     Voter  RaftProtocol
XXX27      ec4c9e05xxx        [xxx]:4647   leader    true   3
XXX28      f57cdac1xxx        [xxx]:4647   follower  true   3
XXX29      cecc6acdxxx        [xxx]:4647   follower  true   3
XXX30      03ee1ddbxxx        [xxx]:4647   follower  true   2
XXX31      c7da621cxxx        [xxx]:4647   follower  true   3
```

If I understand the documentation correctly, this is not supposed to be possible:

> This means that once a cluster has been upgraded with servers all running Raft protocol version 3, it will no longer allow servers running any older Raft protocol versions to be added.

(https://www.nomadproject.io/docs/upgrade/upgrade-specific#upgrading-to-raft-protocol-3)

### Reproduction steps

1. Create cluster with `raft_protocol=3`
2. Remove `raft_protocol` configuration from one node
3. Node rejoins the cluster with `raft_protocol=2`

#### Expected Result

I expected node XXX30 to rejoin the cluster with `raft_protocol=3` even though that option was no longer forced by it's configuration.

#### Actual Result

XXX30 rejoins with `raft_protocol=2`, even though the rest of the cluster is using 3.

### PR DESCRIPTION
**Pull request by lgfa29**
Downgrading the Raft version protocol is not a supported operation.
Checking for a downgrade is hard since this information is not stored in
any persistent place. When a server re-joins a cluster with a prior Raft
version, the Serf tag is updated so Nomad can't tell that the version
changed.

Mixed version clusters must be supported to allow for zero-downtime
rolling upgrades. During this it's expected that the cluster will have
mixed Raft versions. Enforcing consistency strong version consistency
would disrupt this flow.

The approach taken here is to store the Raft version on disk. When the
server starts the `raft_protocol` value is written to the file
`data_dir/raft/version`. If that file already exists, its content is
checked against the current `raft_protocol` value to detect downgrades
and prevent the server from starting.

Any other types of errors are ignore to prevent disruptions that are
outside the control of operators. The only option in cases of an invalid
or corrupt file would be to delete it, making this check useless. So
just overwrite its content with the new version and provide guidance on
how to check that their cluster is an expected state.

Closes #11867

### CODE REVIEW

File: nomad/testing.go [hunk 1]
```diff
@@ -39,6 +39,14 @@ func TestACLServer(t *testing.T, cb func(*Config)) (*Server, *structs.ACLToken,
...
 
 func TestServer(t *testing.T, cb func(*Config)) (*Server, func()) {
+	s, c, err := TestServerWithErr(t, cb)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	return s, c
+}
+
+func TestServerWithErr(t *testing.T, cb func(*Config)) (*Server, func(), error) {
```
Comment (lgfa29) [inline nomad/testing.go:49]: I'm not sure if this is the best approach. I need to test the server doesn't start, but it was always failing the test due to the `t.Fatalf`. 

I found this [other approach](https://blog.antoine-augusti.fr/2015/12/testing-an-os-exit-scenario-in-golang/), but it doesn't sound right for this case.

File: nomad/server.go [hunk 2]
```diff
@@ -1399,6 +1409,42 @@ func (s *Server) setupRaft() error {
...
+// the Raft version is incompatible with the current version configured.
+// Provide best-effort check if the file cannot be read.
+func (s *Server) checkRaftVersionFile(path string) error {
+	raftVersion := s.config.RaftConfig.ProtocolVersion
+	baseWarning := "use the 'nomad operator raft list-peers' command to make sure the Raft protocol versions are consistent"
+
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			s.logger.Info(fmt.Sprintf("Raft version file not found, %v", baseWarning))
```
Comment (tgross) [inline nomad/server.go:1422]: Won't this pop up on first boot for everyone? I might leave this log message off or set it to debug so it's not alarming to first-time users.

File: nomad/server.go [hunk 3]
```diff
@@ -1284,6 +1284,16 @@ func (s *Server) setupRaft() error {
 			return err
 		}
 
+		// Check Raft version and update the version file.
+		raftVersionFilePath := filepath.Join(path, "version")
+		raftVersionFileContent := fmt.Sprintf("%d", s.config.RaftConfig.ProtocolVersion)
+		if err := s.checkRaftVersionFile(raftVersionFilePath); err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(raftVersionFilePath, []byte(raftVersionFileContent), 0755); err != nil {
```
Comment (tgross) [inline nomad/server.go:1293]: I can't think of any reason we'd want this to be executable:

```suggestion
		if err := ioutil.WriteFile(raftVersionFilePath, []byte(raftVersionFileContent), 0644); err != nil {
```

File: nomad/server.go [hunk 4]
```diff
@@ -1399,6 +1409,42 @@ func (s *Server) setupRaft() error {
...
+		return nil
+	}
+
+	v, err := ioutil.ReadFile(path)
+	if err != nil {
+		s.logger.Warn(fmt.Sprintf("unable to read Raft version file, %s", baseWarning), "error", err)
+		return nil
+	}
+
+	previousVersion, err := strconv.Atoi(string(v))
```
Comment (tgross) [inline nomad/server.go:1435]: We should probably trim whitespace on the string contents, just to make sure we're not failing with an extra newline from when someone checked it in their editor and accidentally saved.

File: nomad/testing.go (hunk 1 above)
Comment (tgross) [reply nomad/testing.go:49]: Seems reasonable

File: nomad/server.go [hunk 5]
```diff
@@ -1399,6 +1409,42 @@ func (s *Server) setupRaft() error {
...
+			s.logger.Info(fmt.Sprintf("Raft version file not found, %v", baseWarning))
+		} else {
+			s.logger.Warn(fmt.Sprintf("unable to read Raft version file, %s", baseWarning), "error", err)
+		}
+		return nil
+	}
+
+	v, err := ioutil.ReadFile(path)
+	if err != nil {
+		s.logger.Warn(fmt.Sprintf("unable to read Raft version file, %s", baseWarning), "error", err)
```
Comment (tgross) [inline nomad/server.go:1431]: In this case and the case where we can't stat the file but it exists, is the `baseWarning` the right action to recommend? It seems like we want them to remove the old raft version file too?

**Review by tgross — APPROVED** [review]
LGTM overall, my comments are mostly nitpicking over error messages 😀

File: nomad/testing.go [hunk 6]
```diff
@@ -39,6 +39,14 @@ func TestACLServer(t *testing.T, cb func(*Config)) (*Server, *structs.ACLToken,
 }
 
 func TestServer(t *testing.T, cb func(*Config)) (*Server, func()) {
+	s, c, err := TestServerWithErr(t, cb)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
```
Comment (shoenig) [inline nomad/testing.go:45]:
```suggestion
    require.NoError(t, err) 
```

File: nomad/testing.go (hunk 1 above)
Comment (shoenig) [inline nomad/testing.go:49]: nit: `With` usually means _with parameter_, so how about `TestServerErr`

File: nomad/server_test.go [hunk 7]
```diff
@@ -645,3 +645,29 @@ func TestServer_ReloadSchedulers_InvalidSchedulers(t *testing.T) {
 	currentWC = s.GetSchedulerWorkerConfig()
 	require.Equal(t, origWC, currentWC)
 }
+
+func TestServer_PreventRaftDowngrade(t *testing.T) {
+	ci.Parallel(t)
+
+	dir := tmpDir(t)
+	defer os.RemoveAll(dir)
```
Comment (shoenig) [inline nomad/server_test.go:653]:
```suggestion
	dir := t.TempDir()
```

https://pkg.go.dev/testing#T.TempDir

File: nomad/server.go [hunk 8]
```diff
@@ -1284,6 +1284,16 @@ func (s *Server) setupRaft() error {
 			return err
 		}
 
+		// Check Raft version and update the version file.
+		raftVersionFilePath := filepath.Join(path, "version")
+		raftVersionFileContent := fmt.Sprintf("%d", s.config.RaftConfig.ProtocolVersion)
```
Comment (shoenig) [inline nomad/server.go:1289]:
```suggestion
		raftVersionFileContent := strconv.Itoa(s.config.RaftConfig.ProtocolVersion)
```

**Review by shoenig — APPROVED** [review]
LGTM; just nitpicks

File: nomad/server.go (hunk 3 above)
Comment (lgfa29) [reply nomad/server.go:1293]: Ah thanks, I think it was bad copy-pasta from the [`peers.info` file](https://github.com/hashicorp/nomad/blob/main/nomad/server.go#L1333), which doesn't seem like it's supposed to be an executable either.

File: nomad/testing.go (hunk 1 above)
Comment (lgfa29) [reply nomad/testing.go:49]: Ah yeah, good point. Thanks!

File: nomad/server_test.go (hunk 7 above)
Comment (lgfa29) [reply nomad/server_test.go:653]: Nice, TIL. The `tmpDir` function is used in a bunch a places, so I won't try updating the other calls in this PR, but definitely something to keep in mind.

File: nomad/server.go (hunk 5 above)
Comment (lgfa29) [reply nomad/server.go:1431]: Good question, I'm not sure to be honest. Originally I had it so any error here would bubble up and prevent the server from starting, but that seemed not great.

In the case that we can't read the version file, I didn't return an error to allow the file to be overwritten. If the problem is some kind of hardware failure, that operation will result in an error. If the problem is the file content, we just overwrite it with the new value.

So the goal of `baseWarning` is alert users that we weren't able to verify the previous Raft version, so they should do it themselves. Maybe the `baseWarning` message itself should be more explicit?

Comment (tgross) [reply nomad/server.go:1431]: Ah yeah I see what you mean about how narrowly you scoped when to return an error. So on second thought, I think what you've got here LGTM

File: nomad/server.go (hunk 3 above)
Comment (shoenig) [reply nomad/server.go]: I'll open a PR for that

**Pull request merged**